pkg/server/api/agent/v1/service_test.go

```diff
@@ -1362,13 +1355,234 @@ func TestCreateJoinTokenWithAgentId(t *testing.T) {
+				Step: &agentpb.AttestAgentRequest_Params_{
+					Params: &agentpb.AttestAgentRequest_Params{
+						Data: &types.AttestationData{
+							Type: "foo type",
+						},
+						Params: &agentpb.AgentX509SVIDParams{},
+					},
+				},
+			},
+			err:    "invalid argument: missing CSR",
+			logMsg: "Invalid argument",
+			code:   codes.InvalidArgument,
+		},
+
+		{
+			name:           "rate limit fails",
+			request:        &agentpb.AttestAgentRequest{},
+			err:            "rate limit fails",
+			logMsg:         "Rejecting request due to attest agent rate limiting",
+			rateLimiterErr: status.Error(codes.Unknown, "rate limit fails"),
+			code:           codes.Unknown,
+		},
+
+		{
+			name:    "join token does not exist",
+			request: getAttestAgentRequest("join_token", "bad_token", testCsr),
+			err:     "join token does not exist or has already been used",
+			logMsg:  "Join token does not exist or has already been used",
+			code:    codes.InvalidArgument,
+		},
+
+		{
+			name:     "attest with join token",
+			request:  getAttestAgentRequest("join_token", "test_token", testCsr),
+			response: getAttestAgentResponse("/spire/agent/join_token/test_token"),
+		},
+
+		{
+			name:     "attest with join token is banned",
+			isBanned: true,
+			err:      "failed to attest; agent is banned",
+			logMsg:   "Failed to attest; agent is banned",
+			request:  getAttestAgentRequest("join_token", "test_token", testCsr),
+			response: getAttestAgentResponse("/spire/agent/join_token/test_token"),
+			code:     codes.PermissionDenied,
+		},
+
+		{
+			name:     "attest with result",
+			request:  getAttestAgentRequest("test_type", "test_payload", testCsr),
+			response: getAttestAgentResponse("/spire/agent/test_type/test_agent_id"),
+		},
+
+		{
+			name:         "attest with challenge",
+			hasChallenge: true,
+			request:      getAttestAgentRequest("test_type", "test_payload", testCsr),
+			response:     getAttestAgentResponse("/spire/agent/test_type/test_agent_id"),
+		},
+
+		{
+			name:         "attest banned",
+			hasChallenge: false,
+			isBanned:     true,
+			err:          "failed to attest; agent is banned",
+			logMsg:       "Failed to attest; agent is banned",
+			request:      getAttestAgentRequest("test_type", "test_payload", testCsr),
+			code:         codes.PermissionDenied,
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			test := setupServiceTest(t)
+			defer test.Cleanup()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			test.setupJoinToken(ctx, t)
+			test.setupAttestor(t, tt.hasChallenge)
+			test.setupResolver(t)
+			if tt.isBanned {
+				test.setupBannedNodes(ctx, t)
+			}
+
+			test.rateLimiter.count = 1
+			test.rateLimiter.err = tt.rateLimiterErr
+
+			// create stream
+			stream, err := test.client.AttestAgent(ctx)
+			require.NoError(t, err)
+
+			// send
+			err = stream.Send(tt.request)
+			require.NoError(t, err)
+
+			// recv
+			resp, err := stream.Recv()
```

Comment:
if `err != nil` does not means resp is nil? in that case 
```
			challenge := resp.GetChallenge()
			result := resp.GetResult()
```
will panic

Reply:
can you add a test case where we update an attested node?

Reply:
Go allows calling methods on nil pointers. That's why it hasn't failed so far.